feat(txpool): add configurable batch timeout to transaction pool insertions

keanji-x opened the pull request:
Adds configurable transaction pool insertion batching with `--txpool.batch-timeout`.

The batcher now has two explicit modes. Immediate mode (no timeout, the default) processes requests as soon as they arrive and spawns insertion tasks without a concurrency bound, identical to the previous behavior. Batch-and-timeout mode coalesces requests until either `max-batch-size` is reached or the timeout expires after the first buffered transaction, and bounds in-flight insertion batches via `max_concurrent_batches` (derived from `txpool.additional-validation-tasks + 3`) so submission bursts cannot create unbounded concurrent batches while keeping validation workers saturated.

mattsse commented on crates/transaction-pool/src/batcher.rs:
```diff
@@ -98,23 +170,51 @@ where
+                    }
+                    Poll::Pending => return Poll::Pending,
+                }
+            },
```
all of this looks a bit too complex

how should this behave exactly?

can we not use just the interval to yield pending if it's not ready yet?

keanji-x replied:
Fixed, thanks for review

mattsse reviewed (changes requested):
this makes sense and I#ve thought about this as well because most of the time this ends up just batching 1 single request, but even waiting a bit should be very beneficial here.

left some suggestions and a q re the poll logic

mattsse commented on crates/transaction-pool/src/batcher.rs:
```diff
@@ -99,22 +171,46 @@ where
-                continue;
+            // 1. Collect available requests (non-blocking)
+            while this.buf.len() < *this.max_batch_size {
+                match this.request_rx.as_mut().poll_recv(cx) {
```
I think this should keep using poll_recv_many

mattsse commented on crates/transaction-pool/src/batcher.rs:
```diff
@@ -99,22 +171,46 @@ where
+
+            if batch_full || timeout_ready {
+                Self::spawn_batch(this.pool, this.buf);
+                this.buf.reserve(*this.max_batch_size);
```
this is wasteful, because this will most likely over allocate

mattsse reviewed (changes requested):
I'd like to restructure the poll logic a bit so that this behaves like:

1. poll_recv many
2. if batch empty -> return early
3. if batch full -> spawn + continue
4. if batch not full and interval is set: poll interval


I'd also like to see some manual poll tests for this

keanji-x commented:
Hi @mattsse 👋 Sorry for the delay on this — I've addressed all the feedback from both review rounds and rebased onto latest main. Here's a summary of what's been updated:

1. Introduced a dedicated BatchConfig struct (as suggested)
2. Made `interval` an `Option<Interval>` for zero-cost immediate mode
3. Restructured the poll logic to follow your suggested flow: poll_recv_many → empty check → batch full spawn (skip interval) → interval poll for partial batches
4. Removed the wasteful buf.reserve (using std::mem::take now)
5. Renamed timeout_ready → should_flush for clarity
6. Added manual poll tests for: full batch, partial batch with interval, immediate mode, and channel close flush

Would appreciate another look when you get a chance. Thanks! 🙏

keanji-x commented on crates/transaction-pool/src/batcher.rs:
```diff
@@ -99,22 +171,62 @@ where
+                // Reset interval if present
+                if let Some(mut interval) = this.interval.as_mut().as_pin_mut() {
+                    interval.as_mut().reset();
+                }
```
You're right, removed the reset here.

emmajam commented:
hey @keanji-x do we still want this PR?

mattsse committed:
feat(txpool): add configurable batch timeout to transaction pool insertions

mattsse committed:
fix(txpool): apply batch concurrency cap only in batch-and-timeout mode

Immediate mode (no batch timeout, the default) now spawns insertion
tasks without acquiring a permit, preserving the pre-existing unbounded
behavior for the default eth_sendRawTransaction path. The cap otherwise
caused head-of-line blocking with max_batch_size = 1, where a few slow
insertions could stall all RPC transaction intake.

mattsse committed:
feat(txpool): record histogram of dispatched insertion batch sizes

mattsse committed:
feat(txpool): support batch timeout in DefaultTxPoolValues

Allows downstream nodes to configure a default batch timeout the same
way as the other txpool defaults. The CLI arg is now a plain duration
defaulting to the global value, zero meaning immediate mode.
